public/app/features/alerting/unified/components/rule-editor/QueryWrapper.tsx

```diff
@@ -98,7 +98,18 @@ export const QueryWrapper = ({
     // It's unclear as to why this happens, but we need better visibility on why this happens,
     // so we log when it does, and make the query model datasource UID match the datasource UID
     // We already elsewhere work under the assumption that the datasource settings are fetched from the datasourceUid property
-    queryWithDefaults.datasource.uid = query.datasourceUid;
+
+    // This check is necessary for some few cases where the datasource might be an string instead of an object
+    // see: https://github.com/grafana/grafana/issues/96040 for more context
+    if (typeof queryWithDefaults.datasource === 'object' && Boolean(queryWithDefaults.datasource)) {
+      queryWithDefaults.datasource.uid = query.datasourceUid;
+    } else {
+      // if the datasource is a string, we need to convert it to an object, and populate the fields from the query model
+      queryWithDefaults.datasource = {};
+      queryWithDefaults.datasource.uid = query.datasourceUid;
+      queryWithDefaults.datasource.type = query.model.datasource?.type;
+      queryWithDefaults.datasource.apiVersion = query.model.datasource?.apiVersion;
+    }
```

Comment:
Two thoughts from me:

Should we switch this/explicitly special case the scenario of "datasource is a string"? So:
```tsx
if(typeof queryWithDefaults.datasource === 'string') { 
  // do special things
}
```

And could we simplify the special case logic to just be like this?:
```tsx
queryWithDefaults.datasource = {
  uid: query.datasourceUid,
  type: query.model.datasource?.type,
  apiVersion: query.model.datasource?.apiVersion,
};
```

Reply:
> Should we switch this/explicitly special case the scenario of "datasource is a string"? 

I believe it’s safer to check for the opposite condition, as this approach covers all types that are not objects.